File: clients/dart/amenolink/lib/src/cache.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'cache_watcher.dart';
import 'interfaces.dart' as i;
import 'resource_manager.dart';
import 'shared.dart';

class Cache implements i.Cache {
  @override
  final String group;

  Cache(this.group);

  @override
  CacheWatcher watch() => CacheWatcher(group);

  @override
  Future<T?> get<T>(String key) async {
    final rawValue = await _request('GET', _cacheUrl(key));
    if (rawValue == null) return null;
    return parseData<T>(rawValue) as T?;
  }

  @override
  Future<void> set(String key, dynamic value) async {
    dynamic serializedValue = value;
    try {
      serializedValue = (value as dynamic).toJson();
    } catch (_) {
      serializedValue = value;
    }

    await _request('POST', _cacheUrl(key), data: serializedValue);
  }

  @override
  Future<T> getOrCreate<T>(String key, Future<T> Function() creator) async {
    final cachedValue = await get<T>(key);
    if (cachedValue != null) return cachedValue;

    final createdValue = await creator();
    await set(key, createdValue);
    return createdValue;
  }

  @override
  Future<Map<String, dynamic>> all() async {
    final responseData = await _request('GET', _cacheAllUrl());
    if (responseData is! Map<String, dynamic>) {
      throw AmenoException('Resposta inesperada da API de cache');
    }
    return responseData;
  }

  @override
  Future<void> clear() async {
    await _request('DELETE', _cacheAllUrl());
  }

  @override
  Future<void> delete(String key) async {
    await _request('DELETE', _cacheUrl(key));
  }

  String _cacheUrl(String key) {
    final uri = Uri.parse(
      '${clientSetup.originUrl}/api/cache',
    ).replace(queryParameters: {'groupName': group, 'key': key});
    return uri.toString();
  }

  String _cacheAllUrl() {
    final uri = Uri.parse('${clientSetup.originUrl}/api/cache/all').replace(queryParameters: {'groupName': group});
    return uri.toString();
  }

  Future<dynamic> _request(String method, String url, {dynamic data}) async {
    try {
      final headers = <String, String>{};
      String? body;
      if (data != null) {
        headers['Content-Type'] = 'application/json';
        body = jsonEncode(data);
      }

      final uri = Uri.parse(url);
      http.Response response;

      switch (method.toUpperCase()) {
        case 'GET':
          response = await http.get(uri, headers: headers);
          break;
        case 'POST':
          response = await http.post(uri, headers: headers, body: body);
          break;
        case 'DELETE':
          response = await http.delete(uri, headers: headers, body: body);
          break;
        default:
          throw AmenoException('Método HTTP não suportado: $method');
      }

      if (response.statusCode != 200) {
        throw AmenoException('Status HTTP inesperado: ${response.statusCode}');
      }
      if (response.body.isEmpty) return null;
      return jsonDecode(response.body);
    } catch (exception) {
      if (exception is AmenoException) rethrow;
      throw AmenoException('Erro na operação de cache: $exception');
    }
  }
}

Cache cache(String groupName) {
  resourceManager.caches.add(groupName);
  return Cache(groupName);
}

File: clients/csharp/AmenoLink/Cache.cs
using System.Text.Json;

namespace AmenoLink;

public class Cache(string group) : ICache
{
    public string Group { get; } = group;

    public ICacheWatcher Watch() => new CacheWatcher(Group);

    public async Task<T?> Get<T>(string key)
    {
        string url = BuildCacheUrl(key);
        return await Request<T?>("GET", url);
    }

    public async Task Set<T>(string key, T value)
    {
        string url = BuildCacheUrl(key);
        await Request<object?>("POST", url, value);
    }

    public async Task<T> GetOrCreate<T>(string key, Func<Task<T>> creator)
    {
        var cached = await Get<T>(key);
        if (cached is not null)
            return cached;

        var created = await creator();
        await Set(key, created);
        return created;
    }

    public async Task<T> GetOrCreate<T>(string key, Func<T> creator)
    {
        var cached = await Get<T>(key);
        if (cached is not null)
            return cached;

        var created = creator();
        await Set(key, created);
        return created;
    }

    public async Task<Dictionary<string, object?>> All()
    {
        string url = BuildCacheAllUrl();
        var result = await Request<Dictionary<string, object?>>("GET", url);
        return result ?? [];
    }

    public async Task Clear()
    {
        string url = BuildCacheAllUrl();
        await Request<object?>("DELETE", url);
    }

    public async Task Delete(string key)
    {
        string url = BuildCacheUrl(key);
        await Request<object?>("DELETE", url);
    }

    private string BuildCacheUrl(string key)
    {
        string encodedGroup = Uri.EscapeDataString(Group);
        string encodedKey = Uri.EscapeDataString(key);
        return $"{AmenoLinkClient.Settings.OriginUrl}/api/cache?groupName={encodedGroup}&key={encodedKey}";
    }

    private string BuildCacheAllUrl()
    {
        string encodedGroup = Uri.EscapeDataString(Group);
        return $"{AmenoLinkClient.Settings.OriginUrl}/api/cache/all?groupName={encodedGroup}";
    }

    private static async Task<TResult?> Request<TResult>(string method, string url, object? data = null)
    {
        try
        {
            using var request = new HttpRequestMessage(new HttpMethod(method), url);
            if (data is not null)
            {
                string json = JsonSerializer.Serialize(data, AmenoLinkClient.JsonOptions);
                request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
            }

            using var client = new HttpClient();
            var response = await client.SendAsync(request);

            if (!response.IsSuccessStatusCode)
                throw new AmenoException($"Status HTTP inesperado: {(int)response.StatusCode}");

            string content = await response.Content.ReadAsStringAsync();
            if (string.IsNullOrWhiteSpace(content))
                return default;

            return JsonSerializer.Deserialize<TResult>(content, AmenoLinkClient.JsonOptions);
        }
        catch (AmenoException)
        {
            throw;
        }
        catch (Exception exception)
        {
            throw new AmenoException($"Erro na operação de cache: {exception.Message}", exception);
        }
    }
}
